src/sdk/cohort.ts

```diff
@@ -0,0 +1,77 @@
+import { PublicKey } from '@nucypher/nucypher-core';
+
+import { Porter } from '../characters/porter';
+import { ChecksumAddress } from '../types';
+
+export type Ursula = {
+  readonly checksumAddress: ChecksumAddress;
+  readonly uri: string;
+  readonly encryptingKey: PublicKey;
+};
+
+type CohortParameters = {
+  porterUri: string;
+  threshold: number;
+  shares?: number;
+  include?: ChecksumAddress[];
+  exclude?: ChecksumAddress[];
+};
+
+type CohortJSON = {
+  ursulaAddresses: ChecksumAddress[];
+  threshold: number;
+  porterUri: string;
+};
+
+export class Cohort {
+  private constructor(
+    public readonly ursulaAddresses: ChecksumAddress[],
+    public readonly threshold: number,
+    public readonly porterUri: string
+  ) {}
+
+  public static async create({
+    porterUri,
+    threshold,
+    shares = 0,
+    include = [],
+    exclude = [],
+  }: CohortParameters) {
+    if (shares == 0 && include.length == 0) {
+      throw new TypeError('Shares is 0 and Include is an empty array');
+    }
+    if (shares == 0 && include.length > 0) {
+      shares = include.length;
+    }
+
+    const porter = new Porter(porterUri);
+    const ursulas = await porter.getUrsulas(shares, exclude, include);
+    const ursulaAddresses = ursulas.map((ursula) => ursula.checksumAddress);
```

Comment:
Here, Ursula sampling occurs when the Cohort is created - I think this is too early. The sampling should be done just before an actual `grant(...)`. The issue is the difference in timing between when the ursulas to use for the policy are determined, and when the `grant` is to occur. These should happen sequentially and hand-in-hand, because it is not guaranteed that the sampled Ursulas will be up and running at `grant` time. 

Cohort should perhaps have a function that does this, and not do this at Cohort creation time.

Reply:
not sure about this, we wanted Cohort to represent a specific set of Ursulas, which means the sampling already needs to have happened, but I see what mean about them not being available at grant time.

The idea is that I can create a cohort with the Ursulas that I want, and then re-use that forever

Reply:
Right. right.

In a localized sense I get it, but I am still a little concerned (may/may not be warranted) about centralizing forces on specific nodes if the same "static" Cohort object is just repeatedly reused and passed around i.e. if we provide some sort of "static/globalish" set of "good", "better", "best" option, then the same nodes can be used repeatedly, as opposed to resampling ursulas when needed based on configurations provided in the Cohort object.

Need to think about it some more.

Reply:
We didn't really delve into the sampling/re-sampling sequence in our tdec discussions, so it's reasonable in the interest of fulfilling Milestone 5 to temporarily define Cohorts as a static list of pre-specified node addresses.

However, in the medium-term, something resembling Porter's `include` logic is more collusion-resistant, more redundant, economically fairer, and confers greater customizability to the adopter. 

The sampling mechanism as a whole may need a rethink in the context of tdec, but a core principle of the work token model is not to unnecessarily disadvantage new stakers (since staker population growth mitigates centralization and supports the token price). Even though fees are currently distributed/rebated purely by stake size, concentrating actual fragment-handling work amongst early stakers (via the problem with defaults mentioned here https://github.com/nucypher/tdec/issues/51#issuecomment-1249947113), still foments a imbalance in power (i.e. if you're not on the dominant cohort lists, you have less political power even with the same size stake) and could even hasten a DAO-driven change to said fee model (to address the discordance between work & remuneration).

Reply:
We didn't really delve into the sampling/re-sampling sequence in our tdec discussions, so it's reasonable in the interest of fulfilling Milestone 5 to temporarily define Cohorts as a static list of pre-specified node addresses.

However, in the medium-term, something resembling Porter's `include` logic is more collusion-resistant, more redundant, economically fairer, and confers greater customizability to the adopter. 

The sampling mechanism as a whole may need a rethink in the context of tdec, but a core principle of the work token model is not to unnecessarily disadvantage new stakers (since staker population growth mitigates centralization and supports the token price). Even though fees are currently distributed/rebated purely by stake size, concentrating actual fragment-handling work amongst early stakers (via the problem with defaults mentioned here https://github.com/nucypher/tdec/issues/51#issuecomment-1249947113), still foments a imbalance in power (i.e. if you're not on the dominant cohort lists, you have less political power even with the same size stake) and could even hasten a DAO-driven change to said fee model (to address the discordance between work & remuneration).